### PR TITLE
crush: fix potential class id collision

### PR DESCRIPTION
**Pull request by xiexingguo**
Signed-off-by: xie xingguo <xie.xingguo@zte.com.cn>

### CODE REVIEW

File: src/crush/CrushWrapper.h [hunk 1]
```diff
@@ -482,10 +482,40 @@ class CrushWrapper {
...
+      return 0;
+    }
+
+    int32_t class_id = class_name.rbegin()->first + 1;
+    if (class_id >= 0) {
+      return class_id;
+    }
+
+    // wrapped, pick a random start and do exhaustive search
+    class_id = rand() % (numeric_limits<int32_t>::max() + 1);
```
Comment (tchaikov) [inline src/crush/CrushWrapper.h]:
```
/var/ceph/ceph/src/crush/CrushWrapper.h:497:57: warning: integer overflow in expression [-Woverflow]
     class_id = rand() % (numeric_limits<int32_t>::max() + 1);
                          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~
```

File: src/crush/CrushWrapper.h [hunk 2]
```diff
@@ -482,10 +482,40 @@ class CrushWrapper {
     return 0;
   }
 
+  int32_t _alloc_class_id() {
+
```
Comment (tchaikov) [inline src/crush/CrushWrapper.h]: could you remove the empty lines?

File: src/crush/CrushWrapper.h [hunk 3]
```diff
@@ -482,10 +482,40 @@ class CrushWrapper {
...
+    if (class_name.empty()) {
+      return 0;
+    }
+
```
Comment (tchaikov) [inline src/crush/CrushWrapper.h]: ditto.

File: src/crush/CrushWrapper.h [hunk 4]
```diff
@@ -482,10 +482,40 @@ class CrushWrapper {
...
+    int32_t class_id = class_name.rbegin()->first + 1;
+    if (class_id >= 0) {
+      return class_id;
+    }
+
```
Comment (tchaikov) [inline src/crush/CrushWrapper.h]: and here

File: src/crush/CrushWrapper.h (hunk 2 above)
Comment (tchaikov) [inline src/crush/CrushWrapper.h]: also, would be better if we can move this method into `CrushWrapper.cc`

File: src/crush/CrushWrapper.h [hunk 5]
```diff
@@ -482,10 +482,40 @@ class CrushWrapper {
     return 0;
   }
 
+  int32_t _alloc_class_id() {
```
Comment (tchaikov) [inline src/crush/CrushWrapper.h]: this method can be marked as `const`.

**Comment by xiexingguo** [comment]
> /var/ceph/ceph/src/crush/CrushWrapper.h:497:57: warning: integer overflow in expression [-Woverflow] class_id = rand() % (numeric_limits<int32_t>::max() + 1); ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~

Well, I confess that I have stolen this from https://github.com/ceph/ceph/pull/15654. 
Wouldn't it have the same problem, then...

**Comment by xiexingguo** [comment]
Updated but still not very sure this will be the right cure. @tchaikov  Can you help to confirm? Thanks!

**Comment by xiexingguo** [comment]
> Well, I confess that I have stolen this from #15654.
> Wouldn't it have the same problem, then...

bid is of type int16_t, hence numeric_limits<bid_t>::max() can be safely promoted to int for the + 1 operation.

**Comment by xiexingguo** [comment]
already merged

**Pull request closed without merging**